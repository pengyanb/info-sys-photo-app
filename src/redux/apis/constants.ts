export const ACTION_GET_ALBUMS = "ACTION_GET_ALBUMS";
export const ACTION_GET_ALBUMS_RESPONSE = "ACTION_GET_ALBUMS_RESPONSE";
export const ACTION_GET_ALBUMS_FAILED = "ACTION_GET_ALBUMS_FAILED";

export const ACTION_GET_PHOTOS = "ACTION_GET_PHOTOS";
export const ACTION_GET_PHOTOS_RESPONSE = "ACTION_GET_PHOTOS_RESPONSE";
export const ACTION_GET_PHOTOS_FAILED = "ACTION_GET_PHOTOS_FAILED";