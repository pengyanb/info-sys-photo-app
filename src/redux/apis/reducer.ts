import {
  ACTION_GET_ALBUMS,
  ACTION_GET_ALBUMS_RESPONSE,
  ACTION_GET_ALBUMS_FAILED,
  ACTION_GET_PHOTOS,
  ACTION_GET_PHOTOS_RESPONSE,
  ACTION_GET_PHOTOS_FAILED
} from "./constants";

import {
  IAlbumInfo,
  IPhotoInfo,
  IActionGetAlbums,
  IActionGetAlbumsResponse,
  IActionGetAlbumsFailed,
  IActionGetPhotos,
  IActionGetPhotosResponse,
  IActionGetPhotosFailed
 } from "./actions";

 export interface IApiState {
  albums: IAlbumInfo[];
  albumsError: any;
  photos: IPhotoInfo[];
  photosError: any
 } 

const initialState: IApiState = {
  albums: [],
  albumsError: undefined,
  photos: [],
  photosError: undefined
};

const apiReducer = (state: IApiState = initialState, action: IActionGetAlbums | IActionGetAlbumsResponse | IActionGetAlbumsFailed | IActionGetPhotos | IActionGetPhotosResponse | IActionGetPhotosFailed) => {
  switch (action.type) {
    case ACTION_GET_ALBUMS:
      return {
        ...state,
        albums: [],
        albumsError: undefined,
      };
    case ACTION_GET_ALBUMS_RESPONSE:
      return {
        ...state,
        albums: (action as IActionGetAlbumsResponse).albums,
        albumsError: undefined,
      };
    case ACTION_GET_ALBUMS_FAILED:
      return {
        ...state,
        albums: [],
        albumsError: (action as IActionGetAlbumsFailed).error,
      };
    case ACTION_GET_PHOTOS:
      return {
        ...state,
        photos:[],
        photosError: undefined,
      };
    case ACTION_GET_PHOTOS_RESPONSE:
      return {
        ...state,
        photos: (action as IActionGetPhotosResponse).photos,
        photosError: undefined
      };
    case ACTION_GET_PHOTOS_FAILED:
      return {
        ...state,
        photos: [],
        photosError: (action as IActionGetAlbumsFailed).error
      };
    default:
      return state;
  }
};

export default apiReducer;