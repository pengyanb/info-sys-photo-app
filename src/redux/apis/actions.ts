import { ACTION_GET_ALBUMS, ACTION_GET_ALBUMS_RESPONSE, ACTION_GET_ALBUMS_FAILED, ACTION_GET_PHOTOS, ACTION_GET_PHOTOS_RESPONSE, ACTION_GET_PHOTOS_FAILED } from "./constants";

export interface IAlbumInfo {
  userId: number;
  id: number;
  title: string;
}

export interface IPhotoInfo {
  albumId: number;
  id: number;
  title: string;
  url: string;
  thumbnailUrl: string;
}

export interface IActionGetAlbums {
  type: string;
}

export interface IActionGetAlbumsResponse {
  type: string;
  albums: IAlbumInfo[];
}

export interface IActionGetAlbumsFailed {
  type: string;
  error: any;
}

export interface IActionGetPhotos {
  type: string;
}

export interface IActionGetPhotosResponse {
  type: string;
  photos: IPhotoInfo[];
}

export interface IActionGetPhotosFailed {
  type: string;
  error: any;
}

export function actionGetAlbums(): IActionGetAlbums {
  return {
    type: ACTION_GET_ALBUMS
  }
}

export function actionGetAlbumsResponse(albums: IAlbumInfo[]): IActionGetAlbumsResponse {
  return {
    type: ACTION_GET_ALBUMS_RESPONSE,
    albums
  };
}

export function actionGetAlbumsFailed(error: any): IActionGetAlbumsFailed {
  return {
    type: ACTION_GET_ALBUMS_FAILED,
    error
  }
}

export function actionGetPhotos(): IActionGetPhotos {
  return {
    type: ACTION_GET_PHOTOS
  }
}

export function actionGetPhotosResponse(photos: IPhotoInfo[]): IActionGetPhotosResponse {
  return {
    type: ACTION_GET_PHOTOS_RESPONSE,
    photos
  }
}

export function actionGetPhotosFailed(error: any): IActionGetPhotosFailed {
  return {
    type: ACTION_GET_PHOTOS_FAILED,
    error
  }
}