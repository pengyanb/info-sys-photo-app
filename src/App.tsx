import React, { FC, useEffect, useState } from 'react';
import { connect } from "react-redux";
import { IApiState } from "./redux/apis/reducer";
import { actionGetAlbums, actionGetPhotos, IAlbumInfo, IPhotoInfo } from "./redux/apis/actions";

import { Alert, List, Typography, Modal } from "antd";
interface IApp {
  albums: IAlbumInfo[];
  albumsError: any;
  photos: IPhotoInfo[];
  photosError: any;
  getAlbums: () => void;
  getPhotos: () => void;
}

export const App: FC<IApp> = ({ albums, photos, albumsError, photosError, getAlbums, getPhotos }) => {
  useEffect(() => {
    getAlbums();
    getPhotos();
  }, [getAlbums, getPhotos]);

  const [selectedAlbum, setSelectedAlbum] = useState<IAlbumInfo | undefined>(undefined);
  const [albumPhotos, setAlbumPhotos] = useState<IPhotoInfo[]>([]);
  const [selectedPhoto, setSelectedPhoto] = useState<IPhotoInfo | undefined>(undefined)

  const onAlbumClick = (album: IAlbumInfo) => {
    if (album.id === selectedAlbum?.id) {
      setSelectedAlbum(undefined);
      setAlbumPhotos([]);
      setSelectedPhoto(undefined);
    } else {
      setSelectedAlbum(album);
      const thumbnails: IPhotoInfo[] = [];
      photos.map(photo => {
        if (photo.albumId === album.id) {
          thumbnails.push(photo)
        }
      });
      setAlbumPhotos(thumbnails)
    }
  }

  const onThumnailClick = (photo: IPhotoInfo) => {
    setSelectedPhoto(photo);
  };

  if (albumsError) {
    return <Alert type="error" message={albumsError.toString()} showIcon />;
  } else if (photosError) {
    return <Alert type="error" message={photosError.toString()} showIcon />;
  } else {
    return (<>
      <List
        header={<h3>Album list</h3>}
        bordered
        dataSource={albums}
        renderItem={item => (
          <List.Item
            style={{ cursor: "pointer", flexDirection: "column" }}
          >
            <div onClick={() => {
              onAlbumClick(item);
            }}><Typography.Text>{item.title}</Typography.Text></div>
            { item.id === selectedAlbum?.id ? <div>{
              albumPhotos.map((photo, index) => <span onClick={() => {
                onThumnailClick(photo);
              }} key={index}><img alt="thumnail" src={photo.thumbnailUrl} /></span>)
            }</div> : null
            }
          </List.Item>)}
      />
      <Modal title={selectedPhoto?.title} visible={selectedPhoto !== undefined} onOk={() => setSelectedPhoto(undefined)} onCancel={() => setSelectedPhoto(undefined)} width={800}>
        <img alt="full" src={selectedPhoto?.url} />
      </Modal>
    </>);
  }
}

const mapStateToProps = (state: { api: IApiState }) => ({
  albums: state.api.albums,
  albumsError: state.api.albumsError,
  photos: state.api.photos,
  photosError: state.api.photosError
});

const mapDispatchToProps = (dispatch: any) => ({
  getAlbums: () => dispatch(actionGetAlbums()),
  getPhotos: () => dispatch(actionGetPhotos())
});


export default connect(mapStateToProps, mapDispatchToProps)(App);
