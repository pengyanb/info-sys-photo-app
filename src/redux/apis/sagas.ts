import { call, put, takeLatest, all } from "redux-saga/effects"
import axios from "axios";
import { ACTION_GET_ALBUMS, ACTION_GET_PHOTOS } from "./constants";
import { actionGetAlbumsResponse, actionGetAlbumsFailed, actionGetPhotosResponse, actionGetPhotosFailed } from "./actions";

const CONST_ALBUMS_URL = "http://jsonplaceholder.typicode.com/albums";
const CONST_PHOTOS_URL = "http://jsonplaceholder.typicode.com/photos";

export function* getAlbumsRequest() {
  
  try {
    const result = yield call((axios as any), {
      url: CONST_ALBUMS_URL,
      method: "GET"
    });
    console.log("!!!!getAlbumsRequest", result.data);
    yield put(actionGetAlbumsResponse(result.data));
  } catch (err) {
    yield put(actionGetAlbumsFailed(err));
  }
};

export function* getPhotosRequest() {
  try {
    const result = yield call((axios as any), {
      url: CONST_PHOTOS_URL,
      method: "GET"
    });
    yield put(actionGetPhotosResponse(result.data));
  } catch (err) {
    yield put(actionGetPhotosFailed(err));
  }
}

export function* getAlbumsWatcher() {
  yield takeLatest(ACTION_GET_ALBUMS, getAlbumsRequest);
}

export function* getPhotosWatcher() {
  yield takeLatest(ACTION_GET_PHOTOS, getPhotosRequest);
}

export default function* apiSaga() {
  yield all([getAlbumsWatcher(), getPhotosWatcher()]);
}