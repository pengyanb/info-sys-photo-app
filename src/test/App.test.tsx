import React from "react";
import { shallow, mount } from "enzyme";

import { App } from "../App";
import { IAlbumInfo, IPhotoInfo } from "../redux/apis/actions";

describe("App tests", () => {
  it("snapshot test", () => {
    const getAlbums = jest.fn();
    const getPhotos = jest.fn();

    const album: IAlbumInfo = {
      id: 1,
      userId: 1,
      title: "album"
    };

    const photo: IPhotoInfo = {
      id: 1,
      albumId: 1,
      title: "photo",
      thumbnailUrl: "thumbnailUrl",
      url: "url"
    };
    window.matchMedia = jest.fn();
    const renderedApp = shallow(<App albums={[album]} photos={[photo]} getAlbums={getAlbums} getPhotos={getPhotos} albumsError={undefined} photosError={undefined} />);
    expect(renderedApp).toMatchSnapshot();
  });

  it("click on list item and click on thumbnail", () => {
    const getAlbums = jest.fn();
    const getPhotos = jest.fn();

    const album: IAlbumInfo = {
      id: 1,
      userId: 1,
      title: "album"
    };

    const photo: IPhotoInfo = {
      id: 1,
      albumId: 1,
      title: "photo",
      thumbnailUrl: "thumbnailUrl",
      url: "url"
    };
    (window as any).matchMedia = jest.fn(() => ({
      addListener: jest.fn()
    }));
    const renderedApp = mount(<App albums={[album]} photos={[photo]} getAlbums={getAlbums} getPhotos={getPhotos} albumsError={undefined} photosError={undefined} />);

    const listDiv = renderedApp.find('li.ant-list-item > div');
    expect(listDiv).not.toBeUndefined();
    expect(listDiv.html()).toBe('<div><span class=\"ant-typography\">album</span></div>');

    listDiv.simulate("click");

    const img = renderedApp.find('img[alt="thumnail"]');
    expect(img.html()).toEqual('<img alt="thumnail" src="thumbnailUrl">');

    const imgSpan = renderedApp.find("li.ant-list-item>div>span");
    imgSpan.simulate("click");
    const imgFull = renderedApp.find('img[alt="full"]');
    expect(imgFull).not.toBeUndefined();
  })
});

