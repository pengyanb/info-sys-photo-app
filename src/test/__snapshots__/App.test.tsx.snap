// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App tests snapshot test 1`] = `
<Fragment>
  <List
    bordered={true}
    dataSource={
      Array [
        Object {
          "id": 1,
          "title": "album",
          "userId": 1,
        },
      ]
    }
    header={
      <h3>
        Album list
      </h3>
    }
    renderItem={[Function]}
  />
  <Modal
    confirmLoading={false}
    maskTransitionName="fade"
    okType="primary"
    onCancel={[Function]}
    onOk={[Function]}
    transitionName="zoom"
    visible={false}
    width={800}
  >
    <img
      alt="full"
    />
  </Modal>
</Fragment>
`;
